datafusion/execution/src/object_store.rs

```diff
@@ -79,71 +79,42 @@ impl std::fmt::Display for ObjectStoreUrl {
     }
 }
 
-/// Provides a mechanism for lazy, on-demand creation of an [`ObjectStore`]
-///
-/// For example, to support reading arbitrary buckets from AWS S3
-/// without instantiating an [`ObjectStore`] for each possible bucket
-/// up front, an [`ObjectStoreProvider`] can be used to create the
-/// appropriate [`ObjectStore`] instance on demand.
-///
-/// See [`ObjectStoreRegistry::new_with_provider`]
-pub trait ObjectStoreProvider: Send + Sync + 'static {
-    /// Return an ObjectStore for the provided url, called by [`ObjectStoreRegistry::get_by_url`]
-    /// when no matching store has already been registered. The result will be cached based
-    /// on its schema and authority. Any error will be returned to the caller
+/// Provides a mechanism to get and put object stores.
+pub trait ObjectStoreRegistry: Send + Sync + std::fmt::Debug + 'static {
+    /// If a store with the same schema and host existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        scheme: &str,
+        host: &str,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Insert a [`ObjectStore`] with the key of a given url got by [`get_url_key()`]
+    ///
+    /// If a store with the same url key, it is replaced and returned
+    fn put_with_url(
```

Comment:
Do we need both a `put_with_url` and `register_store`?

The motivation for not passing the entire URL, was to avoid confusion over what bits of the URL "counted", e.g. should credential influence this? I don't feel strongly, but feel we probably shouldn't have two methods that do basically the same thing